feat: improved user auth tokens

mdtro opened the pull request:
Supports https://github.com/getsentry/rfcs/pull/32. 

- Newly created _user auth tokens_ will be prefixed with `sntryu_`. 
- Introduce a custom model manager for `ApiToken` to handle the unique creation logic where we need to hash the token values and store them.
- Use the `token_type` (currently optional) parameter/field on `ApiToken` when creating user auth tokens to let the model do the heavy lifting on generating and hashing the values. This will keep the logic out of views and simplify calls to just `new_token = ApiToken.objects.create(user=user, token_type=AuthTokenType.USER)`.
- I've [changed some behavior](https://github.com/getsentry/sentry/pull/68148/files#diff-e68bf726258b71dbfe6c6a3dcb9a959faba4e9585762067078a38bed5bad4812R36-R37) where we only return the `refreshToken` on applicable token types.

I introduce a "read once" pattern in this PR for the token secrets to prevent leaking of them in logs, exceptions, etc. It works like this when creating a new `ApiToken`:

1. The model manager sets temporary fields `__plaintext_token` and `__plaintext_refresh_token` that store the respective plaintext values for temporary reading.
2. When reading the value through the `_plaintext_token` property on `ApiToken` (notice the single prepended underscore) the string value is returned and `__plaintext_token` is immediately set to `None`.
3. If you attempt to read the `_plaintext_token` property again, an exception will be raised, `PlaintextSecretAlreadyRead`.
    - _I opted to raise an exception here rather than just returning `None` as I feel it's more explicit and will prevent unexpected behavior when we accidentally expect a value to be there._

mdtro committed:
feat: improved user auth tokens

mdtro committed:
custom exception when attempting to read more than once

mdtro committed:
remove duplicate code

mdtro committed:
fix test

mdtro committed:
overload update method

mdtro committed:
ref: revamp read-once tokens

mdtro committed:
test: error on access refresh token on user token

mdtro committed:
only send refresh token if it is not a user auth token

mdtro committed:
handle case when refresh_token is set to None after

mdtro committed:
fix typing

nhsiehgit commented on src/sentry/models/apitoken.py:
```diff
@@ -104,10 +240,16 @@ def get_allowed_origins(self):
 
     def refresh(self, expires_at=None):
+        if self.token_type == AuthTokenType.USER:
+            raise NotImplementedError("User auth tokens do not support refreshing the token")
```
Should this be a NotImplementedError then?
Or should we make this like a `raise NotSupported` instead?

mdtro replied:
Ah, yeah, I think that's a better idea.

The [Python docs](https://docs.python.org/3/library/exceptions.html#NotImplementedError) seem to agree as well.

> It should not be used to indicate that an operator or method is not meant to be supported at all – in that case either leave the operator / method undefined or, if a subclass, set it to [None](https://docs.python.org/3/library/constants.html#None).

nhsiehgit reviewed (commented):
hmmm this looks mostly sane to me?

I feel like i would need to significantly refresh my context here to be fully comfortable to sign off on it 😅

azaslavsky commented on tests/sentry/backup/snapshots/ReleaseTests/test_at_head.pysnap:
```diff
@@ -1035,60 +1034,60 @@ source: tests/sentry/backup/test_releases.py
-    token_last_characters: c9f4
+    token: c9954191fded44bc82f6e924c4224975a833f542e5aaa301ccdd31d588094c69
+    token_last_characters: 4c69
     token_type: null
```
Intentional to have some tokens with no type, right?

mdtro replied:
Yes, that's correct. To start we will only be populating the token_type for User Auth Tokens.

azaslavsky commented on tests/sentry/backup/snapshots/ReleaseTests/test_at_head.pysnap:
```diff
@@ -1035,60 +1034,60 @@ source: tests/sentry/backup/test_releases.py
     expires_at: null
     hashed_refresh_token: null
-    hashed_token: null
+    hashed_token: f1366f7fc8cf21fdc979ca2fabe6a03fad692f0f46f90ca813968e110680066f
```
Is this a possible state? Has a token, but no refresh_token?

mdtro replied:
Correct. User Auth Tokens are standard non-expiring API tokens at the moment. They do not utilize OAuth flows and thus should not have a refresh token. This model serves several token functions in the code base unfortunately and can get confusing.

In a typical OAuth flow a refresh token would be used to get a new access token (that's stored in `token`) when it expires. User auth tokens aren't applicable here.

azaslavsky commented on src/sentry/models/apitoken.py:
```diff
@@ -125,9 +267,9 @@ def write_relocation_import(
         existing = self.__class__.objects.filter(query).first()
         if existing:
-            self.token = generate_token()
+            self.token = generate_token(token_type=self.token_type)
```
👍

azaslavsky reviewed (approved):
Awesome, LGTM from the relocation perspective!

mdtro committed:
use more appropriate exception name

mdtro committed:
test refresh tokens serialization

ykamo001 commented on src/sentry/models/apitoken.py:
```diff
@@ -28,10 +29,98 @@ def default_expiration():
+        kwargs["refresh_token"] = plaintext_refresh_token
+
+        # Store the plaintext tokens for one-time retrieval
+        self.__plaintext_token = plaintext_token
+        self.__plaintext_refresh_token = plaintext_refresh_token
```
This class is supposed to manage logic for all objects, I think storing the values in this class probably doesn't make sense? I could be wrong, and maybe this class gets instantiated per object and maybe this is totally safe, but I think the manager class shouldn't be used like this. I believe it gets instantiated once as part of the model, because it's a class attribute and not an instance attribute

ykamo001 replied:
Based on this, i think the later property methods might have to change, so not going to comment on any of them yet

ykamo001 replied:
![Screenshot 2024-04-10 at 11 49 22 AM](https://github.com/getsentry/sentry/assets/5581484/9e153c2e-3bc7-40ac-8670-206bfb92d84d)
Yeah we probably don't want this, does this example make sense?

ykamo001 replied:
I think you want to add the values on the model object created, so
```
object_created = super().create(*args, **kwargs)
# somehow set these values but only once
object_created.__plaintext_token = plaintext_token
object_created.__plaintext_refresh_token = plaintest_refresh_token
```

mdtro replied:
Very good catch! 🥔

Thank you for pairing with me on this the other day. I'm marking this as resolved since we've moved the logic to the actual model class to avoid this issue.

ykamo001 commented on src/sentry/models/apitoken.py:
```diff
@@ -28,10 +29,98 @@ def default_expiration():
+        self.__plaintext_token = plaintext_token
+        self.__plaintext_refresh_token = plaintext_refresh_token
+
+        return super().create(*args, **kwargs)
```
I think maybe what you're looking for is to store those values as part of the object that gets created from the parent create?

mdtro commented on src/sentry/models/apitoken.py:
```diff
@@ -28,10 +29,98 @@ def default_expiration():
 
 
-def generate_token():
+def generate_token(token_type: AuthTokenType | str | None = AuthTokenType.__empty__) -> str:
```
I was running in to some issues with `mypy` on this when I originally just had the type as `AuthTokenType`. 
The particular field on the class is a `CharField` so it's expecting a `str`.

Is there a way around this?

https://github.com/getsentry/sentry/blob/3a86fa61bf4293c0eed9837f6b0529bda15dfbc4/src/sentry/models/apitoken.py#L135

ykamo001 replied:
It just depends what you want to support in the method, and the caller has to make sure to appease the contract. What is the expected flow or contract we want to have?

mdtro replied:
Ultimately, as we improve all of the various token types, I would like to require `AuthTokenType`. All newly generated tokens would be classified with this and we wouldn't allow a `None`. Right now, the `None` is there for backwards compatibility.

Since we've set `choices` on the `CharField`, then I suspect we will end up with a validation error if someone provides a string that doesn't match one of the choices. 🤔 I'll test that out as we progress through the token types.

mdtro committed:
simplify exception classes

mdtro commented on src/sentry/models/apitoken.py:
```diff
@@ -28,10 +29,98 @@ def default_expiration():
+        self,
+        message="the secret you are trying to read is read-once and cannot be accessed directly again",
+    ):
+        super().__init__(message)
```
Today I learned that this is even valid. :D 

```python
class PlaintextSecretAlreadyRead(Exception):
    """the secret you are trying to read is read-once and cannot be accessed directly again"""
    pass


class NotSupported(Exception):
    """the method you called is not supported by this token type"""
    pass
```

ykamo001 replied:
Because python is whitespace based, you just need something to let python know the class definition has ended; which is why the doc string works just fine haha

mdtro commented on src/sentry/models/apitoken.py:
```diff
@@ -28,10 +29,98 @@ def default_expiration():
+
+        if token_type == AuthTokenType.USER:
+            plaintext_token = generate_token(token_type=AuthTokenType.USER)
+            plaintext_refresh_token = None  # user auth tokens do not have refresh tokens
```
Yes, in this case it is intended. User Auth Tokens do not have refresh tokens and should never have them.

In practice we generally create user auth tokens with `ApiToken.objects.create(user=user, refresh_token=None)` in our views, but my goal is to eventually get us to `ApiToken.objects.create(user=user, token_type=AuthTokenType.USER)` so that all of the heavy lifting is happening in the model.

mdtro committed:
ref: read-once logic on apitoken

- Setting the plaintext values on the manager class is wrong and would
cause issues when creating multiple instances of ApiToken.
  - it results in the plaintext value always being the latest instance
    of ApiToken that was created
  - moving this to the model fixes the issue
- introduce setter functions for the plaintext token values
- add docstrings
- remove leading `_` on property functions that return the the plaintext
token values
- after reading, set the token to string value stored in `TOKEN_REDACTED` so
it can still be printed and we can search for the string in log data
where accidental leaks may happen
  - we still throw `PlaintextSecretAlreadyRead` when attempting to read
    the value more than once

mdtro commented on src/sentry/models/apitoken.py:
```diff
@@ -28,10 +29,98 @@ def default_expiration():
+            # to maintain compatibility with current
+            # code that currently calls create with token= specified
+            if "token" in kwargs:
+                plaintext_token = kwargs["token"]
```
There are tests that set tokens that I wouldn't consider valid. For example, their length is way too short. 
I think it would be wise to at least validate length, so I'll get that added in and update all the tests. 

I don't want to change too much behavior in this PR though, so I'll leave that to a future PR -- if you're good with that.

ykamo001 replied:
Yeah that sounds good to me!

mdtro committed:
mypy fix and test fixes

- update tests to access correct property
- ignore typing error

mdtro committed:
add type hints

mdtro committed:
regenerate backup snapshot

ykamo001 reviewed (approved):
Thanks for making all the changes!! Excited about the new flow!
